### PR TITLE
bugfix: change dtype of confmat to int64

### LINKED ISSUE #708: ConfusionMatrix is float32 which gives slightly varying ground truth counts
## 🐛 Bug

confmat in ConfusionMatrix is of data type float32 (see: https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/classification/confusion_matrix.py#L121). This results in slightly varying ground truth counts when the confusion matrix contains a lot of values (e.g. for semantic segmentation):

### To Reproduce

```python
import torch
from torchmetrics import ConfusionMatrix

batchsize = 256
n_classes = 2
width = 512
height = 512

metric = ConfusionMatrix(num_classes=n_classes)

target = torch.randint(size=(batchsize, height, width), low=0, high=n_classes)

for i in range(10):
    preds = torch.randint(size=(batchsize, height, width), low=0, high=n_classes)
    metric.update(preds=preds, target=target)
    print(torch.sum(metric.confmat, axis=1))
    metric.reset()
```

The output is:

```bash
tensor([33556800., 33552060.])
tensor([33556804., 33552060.])
tensor([33556800., 33552060.])
tensor([33556804., 33552060.])
tensor([33556800., 33552060.])
tensor([33556804., 33552060.])
tensor([33556804., 33552060.])
tensor([33556804., 33552060.])
tensor([33556800., 33552060.])
tensor([33556804., 33552060.])
```

### Expected behavior

The number of ground truth pixels should not vary. When I change the data type of confmat to int, everything is fine.

### Environment

- PyTorch Version (e.g., 1.0): 1.10.0
- OS (e.g., Linux): Linux
- How you installed PyTorch (`conda`, `pip`, source): conda
- Python version: 3.8

### PR DESCRIPTION
**Pull request by mona0809**
Fixes slightly varying ground truth counts in confusion matrix

## What does this PR do?

Fixes #708

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]
mind add also note to changelog :rabbit:

File: torchmetrics/classification/confusion_matrix.py [hunk 1]
```diff
@@ -118,7 +118,10 @@ def __init__(
         if self.normalize not in allowed_normalize:
             raise ValueError(f"Argument average needs to one of the following: {allowed_normalize}")
 
-        default = torch.zeros(num_classes, 2, 2) if multilabel else torch.zeros(num_classes, num_classes)
+        if multilabel:
+            default = torch.zeros(num_classes, 2, 2, dtype=torch.long)
+        else:
+            default = torch.zeros(num_classes, num_classes, dtype=torch.long)
```
Comment (justusschock) [inline torchmetrics/classification/confusion_matrix.py:124]: This won't be sufficient. We also need to make sure that the dtype is also preserved when calling metric.to(`torch.float`). IMO in that case we should keep the internal state like that and only convert the output dtype

**Review by justusschock — CHANGES_REQUESTED** [review]

File: torchmetrics/classification/confusion_matrix.py (hunk 1 above)
Comment (SkafteNicki) [reply torchmetrics/classification/confusion_matrix.py:124]: @justusschock Actually, this will work since PR https://github.com/PyTorchLightning/metrics/pull/493 made sure that `metric.to(torch.float)` will no longer change the underlying data type of the metric state (we saw problems with mixed precision). User have to use `metric.set_dtype(torch.float)` to change the dtype and in that case its on the user that the change the precision.

Comment (Borda) [reply torchmetrics/classification/confusion_matrix.py:124]: but still some tests are failing :]

Comment (Borda) [reply torchmetrics/classification/confusion_matrix.py:124]: fixed :rabbit:

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@justusschock mind re-review?
with respect to https://github.com/PyTorchLightning/metrics/pull/715#discussion_r780705051

**Review by SkafteNicki — APPROVED** [review]

**Comment by Borda** [comment]
@justusschock mind re-review :rabbit:

**Review by justusschock — APPROVED** [review]
